state-chain/chains/src/refund_parameters.rs

```diff
@@ -0,0 +1,255 @@
+
+impl<A, D> ChannelRefundParameters<A, D> {
+	pub fn min_output_amount(&self, input_amount: AssetAmount) -> AssetAmount {
+		use sp_runtime::traits::UniqueSaturatedInto;
+		cf_amm_math::output_amount_ceil(input_amount.into(), self.min_price).unique_saturated_into()
+	}
+}
+
+#[cfg(feature = "std")]
+impl RpcChannelRefundParameters {
+	pub fn parse_refund_address_for_chain(
+		self,
+		chain: cf_primitives::ForeignChain,
+	) -> anyhow::Result<ChannelRefundParametersUncheckedEncoded> {
+		Ok(ChannelRefundParameters {
+			retry_duration: self.retry_duration,
+			refund_address: self.refund_address.try_parse_to_encoded_address(chain)?,
+			min_price: self.min_price,
+			refund_ccm_metadata: self.refund_ccm_metadata.clone(),
+		})
+	}
+}
+
+impl<A, D> ChannelRefundParameters<A, D> {
+	pub fn map_address<B, F: FnOnce(A) -> B>(self, f: F) -> ChannelRefundParameters<B, D> {
+		ChannelRefundParameters {
+			retry_duration: self.retry_duration,
+			refund_address: f(self.refund_address),
+			min_price: self.min_price,
+			refund_ccm_metadata: self.refund_ccm_metadata,
+		}
+	}
+	pub fn try_map_address<B, E, F: FnOnce(A) -> Result<B, E>>(
+		self,
+		f: F,
+	) -> Result<ChannelRefundParameters<B, D>, E> {
+		Ok(ChannelRefundParameters {
+			retry_duration: self.retry_duration,
+			refund_address: f(self.refund_address)?,
+			min_price: self.min_price,
+			refund_ccm_metadata: self.refund_ccm_metadata,
+		})
+	}
+}
+
+impl<A> ChannelRefundParametersUnchecked<A> {
+	/// Converts the refund address into the ForeignChainAddress type.
+	pub fn with_foreign_chain_refund_address<C>(
+		self,
+	) -> ChannelRefundParametersUncheckedInternalAddress
+	where
+		C: Chain<ChainAccount = A>,
+		A: Clone + IntoForeignChainAddress<C>,
+	{
+		self.map_address(|addr| addr.clone().into_foreign_chain_address())
+	}
+
+	pub fn validate(
+		&self,
+		refund_asset: Asset,
+		refund_address_decoded: ForeignChainAddress,
+	) -> Result<(), DispatchError> {
+		self.refund_ccm_metadata
+			.as_ref()
+			.map(|refund_ccm| {
+				CcmValidityChecker::check_and_decode(
+					refund_ccm,
+					refund_asset,
+					refund_address_decoded,
+				)
+			})
+			.transpose()?;
+
+		Ok(())
+	}
+}
+
+impl ChannelRefundParametersUncheckedEncoded {
+	/// Try to convert the refund address into the ForeignChainAddress type.
+	pub fn try_with_foreign_chain_refund_address<C: AddressConverter>(
+		self,
+	) -> Result<ChannelRefundParametersUncheckedInternalAddress, DispatchError> {
+		Ok(self.try_map_address(|addr| {
+			C::try_from_encoded_address(addr.clone()).map_err(|_| "Invalid refund address")
+		})?)
+	}
+}
+
+// NOTE: Currently CCM checking requires a ForeignChainAddress - hence the <ForeignChainAddress>
+// constraint on this impl. We should be able to remove this requirement somehow.
+impl ChannelRefundParametersUnchecked<ForeignChainAddress> {
+	/// Checks the CCM Refund metadata payload and converts the address to the internal
+	pub fn with_checked_metadata(
```

Comment:
Same reasoning as the comment above:

```suggestion
	pub fn check_metadata(
```

Reply:
I wanted to find a somewhat common convention that communicates that something inside the struct is being converted. I agree `with_` is not ideal but imo check_metadata sounds like it just checks it without converting. Maybe something like `convert_into` or `map_into`.